File: src/main/java/com/gitlab/daring/image/command/CommandScript.java
package com.gitlab.daring.image.command;

import com.gitlab.daring.image.event.ValueEvent;
import com.gitlab.daring.image.util.CommonUtils;
import com.gitlab.daring.image.util.VoidCallable;
import org.bytedeco.javacpp.opencv_core.Mat;

import static com.gitlab.daring.image.command.CommandScriptUtils.*;

public class CommandScript {

    public final CommandEnv env = new CommandEnv();
    public final ValueEvent<Exception> errorEvent = new ValueEvent<>();

    volatile ScriptCommand command;

    public CommandScript() {
        setText("");
    }

    public String getText() {
        return command.getScript();
    }

    public ScriptCommand getCommand() {
        return command;
    }

    public void setText(String text) {
        tryRun(() -> command = parseScript(text));
    }

    public void runTask(String task) {
        env.setTask(task);
        tryRun(() -> command.execute(env));
    }

    public Mat runTask(String task, Mat mat) {
        env.setMat(mat != null ? mat.clone() : new Mat());
        runTask(task);
        return env.getMat().clone();
    }

    public void execute() {
        runTask("");
    }

    public Mat runCommand(String cmd, Object... args) {
        tryRun(() -> runScript(env, cmdStr(cmd, args)));
        return env.getMat();
    }

    void tryRun(VoidCallable call) {
        CommonUtils.tryRun(call, errorEvent);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/CommandScript.kt
package com.gitlab.daring.image.command

import com.gitlab.daring.image.command.CommandScriptUtils.commandString
import com.gitlab.daring.image.command.CommandScriptUtils.parseScript
import com.gitlab.daring.image.command.CommandScriptUtils.runScript
import com.gitlab.daring.image.event.ValueEvent
import org.bytedeco.javacpp.opencv_core.Mat

class CommandScript {

    val env = CommandEnv()
    val errorEvent = ValueEvent<Exception>()

    var command = parseScript("")

    var text
        get() = command.script
        set(v) = tryRun { command = parseScript(v) }

    init {
        env.startEvent.addListener { startScript() }
    }

    private fun startScript() {
        command.execute(env)
        env.finishEvent.fire(env)
    }

    fun runTask(task: String) = tryRun {
        env.task = task
        env.startEvent.fire(env)
    }

    fun execute() = runTask("")

    fun runTask(task: String, mat: Mat?): Mat {
        env.mat = mat?.clone() ?: Mat()
        runTask(task)
        return env.mat.clone()
    }

    fun runCommand(cmd: String, vararg args: Any): Mat {
        tryRun { runScript(env, commandString(cmd, *args)) }
        return env.mat
    }

    fun tryRun(f: () -> Unit) {
        try { f.invoke() } catch (e: Exception) { errorEvent.fire(e) }
    }

}